### PR TITLE
8196301: java/awt/print/PrinterJob/Margins.java times out

### PR DESCRIPTION
**Pull request by prsadhuk**
This test was timing out in windows in mach5 nightly testing. Investigation reveals that 70% of the time, it is failing due to printer being chosen was Microsoft Print to PDF which opens up a File Save Dialog when "OK" was clicked in printer pagedialog. Since no user intervention is done to dismiss the modal filedialog, so subsequent pagedialog was not dismissed resulting in timeout.
Other times, it was found that "One Note" software printer was used as default printerservice which again, opens up a OneNote app which again gets focus and obscure pagedialog so PageDialog did not get dismissed.
Updated test to ignore "Print To PDF", "OneNote" and "XPS Document Writer"(which again opens filedialog like PDF printer) and run it for several iteration in mach5 platforms which is ok. Link in JBS.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8196301](https://bugs.openjdk.java.net/browse/JDK-8196301): java/awt/print/PrinterJob/Margins.java times out

### Reviewers
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2598/head:pull/2598`
`$ git checkout pull/2598`

### CODE REVIEW

**Comment by azuev-java** [comment]
Shouldn't you update copyright years in the header of the test?

**Comment by prsadhuk** [comment]
No, it's not mandatory to update the copyright year as there is script that should update it at end of release.

**Review by prrace — COMMENTED** [review]
The problem I have with this is that it does not scale.
It is a maintenance nightmare every possible "file printer" on Windows to every test.
and back porting it too ?
removing or disabling one note and the rest s supposed to be part of the set up of the test systems.

And when you say "70%" what is the other 30 % ?

**Comment by prsadhuk** [comment]
other 30% is mixture of OneNote, XPS being used as printer service and timing issue which is why I moved the Thread start closer to actual pagedialog invocation.

**Comment by prsadhuk** [comment]
Maybe @key printer is not being taken seriously during system setup which is why no real printer is being used.

**Comment by prrace** [comment]
> Maybe @key printer is not being taken seriously during system setup which is why no real printer is being used.

That should be irrelevant. jtreg should only be invoking tests which do NOT have the printer keyword since the test job has this : keywords=headful & !printer which are handed directly to jtreg. In other words it is not filtering on whether there is actually a printer, but on whether the keyword is present.

So if this test is being invoked on tests which specify that then there's something badly wrong with jtreg or the harness is mangling it.

But I think this bug was filed back when this test was MISSING that keyword - it was added here 
http://hg.openjdk.java.net/jdk/client/rev/f13dba72a5ea

So likely you should be focused on the timing issue and not the file printers

**Comment by prsadhuk** [comment]
As I told, I already fixed the timing issue too by moving thread start to pass ENTER keypress closer to actual pagedialog invocation and the mach5 job where all the tests passed is there in JBS.

**Comment by prsadhuk** [comment]
> That should be irrelevant. jtreg should only be invoking tests which do NOT have the printer keyword since the test job has this : keywords=headful & !printer which are handed directly to jtreg. In other words it is not filtering on whether there is actually a printer, but on whether the keyword is present.
> 
> So if this test is being invoked on tests which specify that then there's something badly wrong with jtreg or the harness is mangling it.

I have to change script to have keywords=headful & printer as it was not running this test in mach5 without it.

**Comment by prsadhuk** [comment]
Since this is one of the few printer test which is automated, it will fail locally too if someone has "Print to PDF", "XPS" or "OneNote" printer as default since it will open up filedialog or app which will prevent the focus to be on "OK" button of pagedialog, so the fix is applicable locally too, so I would like this fix to be considered.

**Comment by prrace** [comment]
> Since this is one of the few printer test which is automated, it will fail locally too if someone has "Print to PDF", "XPS" or "OneNote" printer as default since it will open up filedialog or app which will prevent the focus to be on "OK" button of pagedialog, so the fix is applicable locally too, so I would like this fix to be considered.

There are other automated print tests. I know I used to spit out a bunch of pages just doing PIT. I think this is the wrong approach. 
@requires should not be confused with keywords.
The checks for PDF printers are not appropriate.

**Review by prrace — APPROVED** [review]

File: test/jdk/java/awt/print/PrinterJob/Margins.java [hunk 1]
```diff
@@ -47,28 +49,33 @@
...
         double hgt = paper.getHeight();
         paper.setImageableArea(0, -10, wid, hgt);
-        t.start();
+
+        Thread t1 = new Thread(() -> {
+            robot.delay(2000);
+            robot.keyPress(KeyEvent.VK_ENTER);
+            robot.keyRelease(KeyEvent.VK_ENTER);
+        });
+        t1.start();
```
Comment (mrserb) [inline test/jdk/java/awt/print/PrinterJob/Margins.java:71]: Don't you need to wait until the t1(and others) thread completes? You call start() but the actual execution may not start till the end of the test.

**Pull request closed without merging**